extension/app/background.ts

```diff
@@ -370,6 +370,99 @@ chrome.runtime.onMessage.addListener(
   }
 );
 
+/**
+ * Listener for messages sent from external websites that are whitelisted on the manifest.
+ * It allows to open the side panel and navigate to a specific conversation.
+ *
+ * We return true to keep the message channel open for async response.
+ */
+chrome.runtime.onMessageExternal.addListener((request) => {
+  if (
+    request.action !== "openSidePanel" ||
+    !request.conversationId ||
+    !request.workspaceId ||
+    !/^[a-zA-Z0-9_-]{10,}$/.test(request.conversationId) ||
+    !/^[a-zA-Z0-9_-]{10,}$/.test(request.workspaceId)
+  ) {
+    log("[onMessageExternal] Invalid params:", request);
+    return true;
+  }
+
+  chrome.tabs.query({ active: true, currentWindow: true }, (tabs) => {
+    if (tabs[0]) {
+      void chrome.sidePanel
+        .open({
+          windowId: tabs[0].windowId,
+        })
+        .then(() => {
+          chrome.storage.local.get(
+            ["extensionReady", "user"],
+            ({ extensionReady, user }) => {
+              if (request.workspaceId != user?.selectedWorkspace) {
+                log("[onMessageExternal] User selected another workspace.");
+                return;
+              }
+
+              const sendMessage = () => {
+                const params = JSON.stringify({
+                  conversationId: request.conversationId,
+                });
+                void chrome.runtime.sendMessage({
+                  type: "EXT_ROUTE_CHANGE",
+                  pathname: "/run",
+                  search: `?${params}`,
+                });
+              };
+
+              if (!extensionReady) {
+                let retries = 0;
```

Comment:
you can also use set state.lastHandler , which will be called as soon as the input bar is visible (so extension opened and logged in)

Reply:
fair, but it would be better to have one single way to handle this. but we can refactor this later

Reply:
Yes creating a card assigned to me for the refactor!